File: 2019/knowit/15/15.js
"use strict";

const fs = require("fs");
const R = require("ramda");

const x = 0;
const y = 1;
const z = 2;

const sideLength = (p1, p2) =>
	Math.sqrt((p1[x] - p2[x])**2 + (p1[y] - p2[y])**2 + (p1[z] - p2[z])**2);

const triangleArea = ([pa, pb, pc]) => {
	// https://en.wikipedia.org/wiki/Herons_formula
	const a = sideLength(pa, pb);
	const b = sideLength(pb, pc);
	const c = sideLength(pc, pa);
	const s = (a + b + c) / 2;

	return Math.sqrt(s*(s-a)*(s-b)*(s-c));
};

R.compose(
	console.log.bind(console),
	tris => {
		// Get the footprint of the model, in mm^2
		const footprint = R.compose(
			R.sum,
			R.map(triangleArea),
			R.filter(R.all(p => p[z] === 0))
		)(tris);

		// Get the height of the model, in mm
		const height = R.compose(
			R.reduce(Math.max, -Infinity),
			R.flatten,
			R.map(R.map(p => p[z]))
		)(tris);

		// Calculate the volume. Divide by 1000 to convert from mm^3 to ml
		return (footprint * height) / 1000;
	},
	R.map(R.splitEvery(3)),
	R.map(R.map(Number)),
	R.map(R.split(",")),
	R.split("\n"),
	R.trim,
)(fs.readFileSync("input.txt", "utf-8"));

File: 2019/knowit/15/15.hs
import System.IO (readFile)
import Data.List.Split

type Point = (Double, Double, Double)
type Triangle = (Point, Point, Point)
type Model = [Triangle]

a :: [[(Int)]] -> [[(Int)]]
a = id

sideLength :: Point -> Point -> Double
sideLength (a1, b1, c1) (a2, b2, c2) =
    sqrt ((a1 - a2)^2 + (b1 - b2)^2 + (c1 - c2)^2)

triangleArea :: Triangle -> Double
triangleArea (pa, pb, pc) = let a = sideLength pa pb
                                b = sideLength pb pc
                                c = sideLength pc pa
                                s = (a + b + c) / 2
                            in  sqrt (s*(s-a)*(s-b)*(s-c))

csvToModel :: String -> Model
csvToModel =
    map (\([a, b, c]) -> (a, b, c))
        . map (map (\([a, b, c]) -> (read a, read b, read c)))
        . map (chunksOf 3)
        . map (splitOn ",")
        . lines

getFootprint :: Model -> Double
getFootprint = 
    foldr (\t a -> a + triangleArea t) 0
        . filter (\((_, _, z1), (_, _, z2), (_, _, z3)) -> z1 == 0 && z2 == 0 && z3 == 0)

getHeight :: Model -> Double
getHeight = 
    maximum
        . map (\((_, _, z1), (_, _, z2), (_, _, z3)) -> maximum [z1, z2, z3])

main =
    print
        . (\m -> let footprint = getFootprint m
                     height    = getHeight m
                 in  (footprint * height) / 1000)
        . csvToModel
        =<< readFile "input.txt"
